rubix-presto/src/main/java/com/qubole/rubix/presto/PrestoClusterManager.java

```diff
@@ -12,149 +12,170 @@
+        ExecutorService executor = Executors.newSingleThreadExecutor();
+        nodesCache = CacheBuilder.newBuilder()
+                .refreshAfterWrite(getNodeRefreshTime(), TimeUnit.SECONDS)
+                .build(CacheLoader.asyncReloading(new CacheLoader<String, List<String>>()
+                {
+                    @Override
+                    public List<String> load(String s)
+                            throws Exception
+                    {
+                        if (!isMaster) {
+                            // First time all nodes start assuming themselves as master and down the line figure out their role
+                            // Next time onwards, only master will be fetching the list of nodes
+                            return ImmutableList.of();
                         }
-                        else {
-                            allNodes = (List<Stats>) allNodesResponse.getValue();
-                        }
-                    }
-                    else {
-                        log.info(String.format("v1/node failed with code: %d setting this node as worker ", allNodesResponse.getStatusCode()));
-                        isMaster = false;
-                        return ImmutableList.of();
-                    }
 
-                    // check on failed nodes
-                    FullJsonResponseHandler.JsonResponse failedNodesResponse = (FullJsonResponseHandler.JsonResponse) failedNodesFuture.get();
-                    if (failedNodesResponse.getStatusCode() == HttpStatus.OK.code()) {
-                        if (!failedNodesResponse.hasValue()) {
-                            failedNodes = ImmutableList.of();
-                            //return ImmutableList.of(InetAddress.getLocalHost().getHostName());
+                        try {
+                            URL allNodesRequest = getNodeUrl();
+                            URL failedNodesRequest = getFailedNodeUrl();
+
+                            HttpURLConnection allHttpCon = (HttpURLConnection) allNodesRequest.openConnection();
+                            allHttpCon.setConnectTimeout(500); //ms
+                            allHttpCon.setRequestMethod("GET");
+
+                            int allNodesResponseCode = allHttpCon.getResponseCode();
+
+                            StringBuffer allResponse = new StringBuffer();
+                            StringBuffer failedResponse = new StringBuffer();
+
+                            if (allNodesResponseCode == HttpURLConnection.HTTP_OK) {
+                                isMaster = true;
+                                BufferedReader in = new BufferedReader(new InputStreamReader(allHttpCon.getInputStream()));
+                                String inputLine = "";
+                                while ((inputLine = in.readLine()) != null) {
+                                    allResponse.append(inputLine);
+                                }
+                                in.close();
```

Comment:
we might leak connections here. If BuffereReader throws exception, we would not close allHttpCon. Same for next request too.
Either refactor so that we can add a finally clause or catch exceptions everywhere appropriately and close required objects.